lang/py/src/avro/avro_builder.py

```diff
@@ -0,0 +1,249 @@
+        return 'double'
+
+    def create_bytes(self):
+        return 'bytes'
+
+    def create_string(self):
+        return 'string'
+
+    def begin_enum(self, name, symbols, namespace=None, aliases=None,
+                   doc=None, **metadata):
+        enum_schema = {
+            'type': 'enum',
+            'name': name,
+            'symbols': symbols
+        }
+        if namespace:
+            self._set_namespace(enum_schema, namespace)
+        if aliases:
+            self._set_aliases(enum_schema, aliases)
+        if doc:
+            self._set_doc(enum_schema, doc)
+        enum_schema.update(metadata)
+
+        self._save_current_schema()
+        self._set_current_schema(enum_schema)
+        return self
+
+    def begin_fixed(self, name, size, namespace=None, aliases=None,
+                    **metadata):
+        fixed_schema = {
+            'type': 'fixed',
+            'name': name,
+            'size': size
+        }
+        if namespace:
+            self._set_namespace(fixed_schema, namespace)
+        if aliases:
+            self._set_aliases(fixed_schema, aliases)
+        fixed_schema.update(metadata)
+
+        self._save_current_schema()
+        self._set_current_schema(fixed_schema)
+        return self
+
+    def begin_array(self, items_schema, **metadata):
+        array_schema = {'type': 'array', 'items': items_schema}
+        array_schema.update(metadata)
+        self._save_current_schema()
+        self._set_current_schema(array_schema)
+        return self
+
+    def begin_map(self, values_schema, **metadata):
+        map_schema = {'type': 'map', 'values': values_schema}
+        map_schema.update(metadata)
+        self._save_current_schema()
+        self._set_current_schema(map_schema)
+        return self
+
+    def begin_record(self, name, namespace=None, aliases=None, doc=None,
+                     **metadata):
+        record_schema = {'type': 'record', 'name': name, 'fields': []}
+        if namespace is not None:
+            self._set_namespace(record_schema, namespace)
+        if aliases:
+            self._set_aliases(record_schema, aliases)
+        if doc:
+            self._set_doc(record_schema, doc)
+        record_schema.update(metadata)
+
+        self._save_current_schema()
+        self._set_current_schema(record_schema)
+        return self
+
+    def begin_field(self, name, typ, has_default=False, default_value=None,
+                    sort_order=None, aliases=None, doc=None, **metadata):
+        field = {'name': name, 'type': typ}
+        if has_default:
+            field['default'] = default_value
+        if sort_order:
+            field['order'] = sort_order
+        if aliases:
+            self._set_aliases(field, aliases)
+        if doc:
+            self._set_doc(field, doc)
+        field.update(metadata)
+
+        self._save_current_schema()
+        self._set_current_schema(field)
+        return self
+
+    def end_field(self):
```

Comment:
Good point. It's originally designed to set field type in a separate function. Since setting the type is now in the same function as creating a field, there is little purpose to have an `end_field` function for it.

By removing the `end_field` function, the `begin_field` will also be renamed to `add_field` since there is no pair functions anymore.